### ISSUE TITLE
File upload progress callback not called evenly nor in the right timeframe

### ISSUE THREAD

**Issue by marianoo-andres** [sendbird/Sendbird-Flutter#10]
Hi,

The progress callback used to listen for file uploading progress is not called evenly nor in the right timeframe. It jumps from 0.003% to 99.999% in 1 millisecond.

### Where it was tested
Mac OS X Catalina
iOS Simulator with iOS 13.7
Flutter 2.0.4
Dart 2.12.2

Flutter doctor output:
```
[✓] Flutter (Channel stable, 2.0.4, on Mac OS X 10.15.6 19G2021 darwin-x64, locale en-AR)
    • Flutter version 2.0.4 at /Users/marianoandres/Flutter/flutter
    • Framework revision b1395592de (4 weeks ago), 2021-04-01 14:25:01 -0700
    • Engine revision 2dce47073a
    • Dart version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/marianoandres/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.7, Build version 11E801a
    ! Xcode 11.7.0 out of date (12.0.1 is recommended).
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 49.0.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] Connected device (2 available)
    • iPhone 8 (mobile) • 4B38CE84-3B3A-42F1-A4B2-DB40E4D19247 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-13-7 (simulator)
    • Chrome (web)      • chrome                               • web-javascript • Google Chrome 90.0.4430.85

! Doctor found issues in 1 category.
```

### Steps to reproduce:

1) Add the progress callback to onSendFileMessage on channel_view_model.dart
```
void onSendFileMessage(File file) async {
    if (file == null) return;

    final params = FileMessageParams.withFile(file);
    final preMessage =
        channel.sendFileMessage(params, onCompleted: (msg, error) {
      final index =
          _messages.indexWhere((element) => element.requestId == msg.requestId);
      if (index != -1) _messages.removeAt(index);
      _messages = [msg, ..._messages];
      _messages.sort((a, b) => b.createdAt.compareTo(a.createdAt));
      markAsReadDebounce();
      if (!isDisposed) notifyListeners();
    },
    progress: (sentBytes, totalBytes) {
      final percent = sentBytes / totalBytes;
      DateTime _now = DateTime.now();
      print("Progress $sentBytes of $totalBytes (${percent * 100}%) Timestamp: ${_now.hour}:${_now.minute}:${_now.second}.${_now.millisecond}");
    });

    _messages = [preMessage, ..._messages];
    if (!isDisposed) notifyListeners();

    lstController.animateTo(
      0.0,
      duration: Duration(milliseconds: 300),
      curve: Curves.easeOut,
    );
  }
```

2) Upload an image or video (to upload a video see 2.video below) and check the prints on the terminal. You should see something similar to the examples below

2.video) First change getImage to getVideo on attachment_modal.dart. Then you will be able to upload video from gallery.
```
Future<File> _showPicker(ImageSource source) async {
    final picker = ImagePicker();
    final pickedFile = await picker.getImage(source: source);
    if (pickedFile != null) {
      return File(pickedFile.path);
    }
    return null;
  }
```

## Examples
In the examples below, I print the bytesSent, totalBytes, progress percentage and the timestamp when the calback is called.

### Example with image
Image uploaded:
https://we.tl/t-8VExSHyzGd

Console output:
```
flutter: Progress 74 of 3782600 (0.0019563263363823824%) Timestamp: 22:59:9.673
flutter: Progress 128 of 3782600 (0.003383915825093851%) Timestamp: 22:59:9.674
flutter: Progress 133 of 3782600 (0.0035161000370115794%) Timestamp: 22:59:9.675
flutter: Progress 135 of 3782600 (0.0035689737217786707%) Timestamp: 22:59:9.677
flutter: Progress 209 of 3782600 (0.005525300058161054%) Timestamp: 22:59:9.679
flutter: Progress 262 of 3782600 (0.0069264527044889754%) Timestamp: 22:59:9.680
flutter: Progress 298 of 3782600 (0.00787817903029662%) Timestamp: 22:59:9.681
flutter: Progress 300 of 3782600 (0.007931052715063714%) Timestamp: 22:59:9.682
flutter: Progress 374 of 3782600 (0.009887379051446096%) Timestamp: 22:59:9.682
flutter: Progress 467 of 3782600 (0.012346005393115846%) Timestamp: 22:59:9.684
flutter: Progress 3782522 of 3782600 (99.99793792629407%) Timestamp: 22:59:9.685
flutter: Progress 3782524 of 3782600 (99.99799079997885%) Timestamp: 22:59:32.225
flutter: Progress 3782600 of 3782600 (100.0%) Timestamp: 22:59:32.226
```

### Example with video
Video uploaded:
https://we.tl/t-dBN3Rvd1TY

Console output:
```
flutter: Progress 74 of 15152360 (0.0004883727683344377%) Timestamp: 22:43:13.107
flutter: Progress 128 of 15152360 (0.0008447528965784867%) Timestamp: 22:43:13.109
flutter: Progress 133 of 15152360 (0.0008777510566010838%) Timestamp: 22:43:13.109
flutter: Progress 135 of 15152360 (0.0008909503206101227%) Timestamp: 22:43:13.111
flutter: Progress 209 of 15152360 (0.0013793230889445603%) Timestamp: 22:43:13.113
flutter: Progress 262 of 15152360 (0.0017291035851840903%) Timestamp: 22:43:13.116
flutter: Progress 298 of 15152360 (0.0019666903373467895%) Timestamp: 22:43:13.117
flutter: Progress 300 of 15152360 (0.0019798896013558283%) Timestamp: 22:43:13.118
flutter: Progress 374 of 15152360 (0.0024682623696902664%) Timestamp: 22:43:13.119
flutter: Progress 472 of 15152360 (0.0031150263061331704%) Timestamp: 22:43:13.120
flutter: Progress 15152282 of 15152360 (99.99948522870365%) Timestamp: 22:43:13.121
flutter: Progress 15152284 of 15152360 (99.99949842796767%) Timestamp: 22:44:23.207
flutter: Progress 15152360 of 15152360 (100.0%) Timestamp: 22:44:23.207
```

### Last comments
As you can see, the result it's pretty similar on both image and video. The progress callback is called within the first milliseconds several times and there is a bump from 0.003% progress to 99.999%. Then, it stops from being called until the file upload finishes.

This has been tested with several images and videos and the behaviour is the same. So, I discard that it has something to do with the file size. Also, this didn't happen to us when we implemented the file upload feature using Sendbird iOS SDK. It seems to me that it's a bug related to this new Sendbird Flutter plugin.

Any thoughts?

Thank you.

**Comment by intoxicated** [comment]
@marianoo-andres Thank you for detail explanation.

First of all, flutter SDK is nothing to do with neither Sendbird SDK for Android nor Sendbird SDK for iOS because it is written in Dart and not using native plugin internally.

I will investigate into this later this week. Thanks again for the report :)

**Comment by intoxicated** [comment]
Not really sure its our problem but here similar problem on SO https://stackoverflow.com/questions/61635094/flutter-how-to-increase-onsendprogress-calls-for-a-fluid-progress-bar-animation 

internally we use `MultipartFile.fromBytes` which may cause the problem..

**Comment by marianoo-andres** [comment]
Yes, it seems it's a problem with the dio library.

According to https://stackoverflow.com/a/67100750/6950858 using `MultipartFile.fromFile` instead of `MultipartFile.fromBytes` would solve the issue but I suppose that would need the file to be saved to a temporary directory.

**Comment by intoxicated** [comment]
We are not using `dio` but `http` package. But yea I think I found a solution 🕺

**Comment by intoxicated** [comment]
FYI for further questions, if you face any problems with sdk problem and not example, please refer https://github.com/sendbird/sendbird-sdk-flutter repo :)

**Comment by marianoo-andres** [comment]
Thanks, I didn't know you had gone live I thought you were still on beta. I will fill report on that repo next time.

**Comment by intoxicated** [comment]
SDK has been GA since 4/22, you can also check out our official docs at https://sendbird.com/docs/chat/v3/flutter/getting-started/about-chat-sdk

**Comment by intoxicated** [comment]
@marianoo-andres the fix has been released with 3.0.12, feel free to reopen if you still face the problem.
